### PR TITLE
f/Start/Stop DB Instance Resource

### LINKED ISSUE #29757: [Enhancement]: Stopping an Amazon RDS DB instance temporarily
### Description

RDS has a capability that the terraform module does not cover - the ability to temporarily stop an RDS cluster.

### Affected Resource(s) and/or Data Source(s)

* aws_db_instance

### Potential Terraform Configuration

```terraform
resource "aws_db_instance" "example" {
  # ... other configuration ...

  stop_temporarily = true
}
```

### References

https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_StopInstance.html

### Would you like to implement a fix?

No

### PR DESCRIPTION
**Pull request by nam054**
### Description

New resource to start and stop RDS DB Instance. This follows the previous established pattern with the EC2 Instance resource and the EC2 Instance State resource.

### Relations

Closes #29757

### Output from Acceptance Testing

```console
> make testacc TESTARGS='-run=TestAccRDSInstanceState_' PKG=rds
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.23.2 test ./internal/service/rds/... -v -count 1 -parallel 20  -run=TestAccRDSInstanceState_ -timeout 360m
2024/11/18 23:59:11 Initializing Terraform AWS Provider...
=== RUN   TestAccRDSInstanceState_basic
=== PAUSE TestAccRDSInstanceState_basic
=== RUN   TestAccRDSInstanceState_update
=== PAUSE TestAccRDSInstanceState_update
=== RUN   TestAccRDSInstanceState_disappears_Instance
=== PAUSE TestAccRDSInstanceState_disappears_Instance
=== CONT  TestAccRDSInstanceState_basic
=== CONT  TestAccRDSInstanceState_disappears_Instance
=== CONT  TestAccRDSInstanceState_update
--- PASS: TestAccRDSInstanceState_basic (542.03s)
--- PASS: TestAccRDSInstanceState_disappears_Instance (542.69s)
--- PASS: TestAccRDSInstanceState_update (960.12s)
PASS

...
```

### CODE REVIEW

File: internal/service/rds/instance.go [hunk 1]
```diff
@@ -2821,6 +2822,41 @@ func waitDBInstanceAvailable(ctx context.Context, conn *rds.Client, id string, t
...
+			instanceStatusStarting,
+			instanceStatusStopping,
+			instanceStatusStorageFull,
+			instanceStatusUpgrading,
+		},
+		Target:     []string{instanceStatusStopped},
+		Refresh:    statusDBInstance(ctx, conn, id),
+		Timeout:    timeout,
+		Delay:      10 * time.Second,
+		MinTimeout: 3 * time.Second,
```
Comment (YakDriver) [inline internal/service/rds/instance.go:2848]:
```suggestion
		Target:                    []string{instanceStatusStopped},
		Refresh:                   statusDBInstance(ctx, conn, id),
		Timeout:                   timeout,
		ContinuousTargetOccurence: 2,
		Delay:                     10 * time.Second,
		MinTimeout:                3 * time.Second,
```

Might consider ContinousTargetOc... Sometimes the only thing that avoids eventual consistency problems.

**Review by YakDriver — DISMISSED** [review]
This is cool. Some say it breaks declarative but if people find it useful, I say, 🎉 

```console
% make testacc TESTARGS='-run=TestAccRDSInstanceState_' PKG=rds
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.23.2 test ./internal/service/rds/... -v -count 1 -parallel 20  -run=TestAccRDSInstanceState_ -timeout 360m
2024/11/19 18:53:05 Initializing Terraform AWS Provider...
=== RUN   TestAccRDSInstanceState_basic
=== PAUSE TestAccRDSInstanceState_basic
=== RUN   TestAccRDSInstanceState_update
=== PAUSE TestAccRDSInstanceState_update
=== RUN   TestAccRDSInstanceState_disappears_Instance
=== PAUSE TestAccRDSInstanceState_disappears_Instance
=== CONT  TestAccRDSInstanceState_basic
=== CONT  TestAccRDSInstanceState_disappears_Instance
=== CONT  TestAccRDSInstanceState_update
--- PASS: TestAccRDSInstanceState_basic (554.67s)
--- PASS: TestAccRDSInstanceState_disappears_Instance (627.12s)
--- PASS: TestAccRDSInstanceState_update (1096.34s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/rds	1100.377s
```

File: website/docs/r/rds_instance_state.html.markdown [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+This resource exports the following attributes in addition to the arguments above:
+
+* `identifier` - DB Instance Identifier
+
+## Timeouts
+
+[Configuration options](https://developer.hashicorp.com/terraform/language/resources/syntax#operation-timeouts):
+
+* `create` - (Default `60m`)
+* `update` - (Default `60m`)
```
Comment (jar-b) [inline website/docs/r/rds_instance_state.html.markdown:42]: To match the resource configuration:
```suggestion
* `create` - (Default `30m`)
* `update` - (Default `30m`)
```

File: website/docs/r/rds_instance_state.html.markdown [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+layout: "aws"
+page_title: "AWS: aws_rds_instance_state"
+description: |-
+  Terraform resource for managing an AWS RDS (Relational Database) RDS Instance State.
+---
+
+# Resource: aws_rds_instance_state
+
+Terraform resource for managing an AWS RDS (Relational Database) RDS Instance State.
+
```
Comment (jar-b) [inline website/docs/r/rds_instance_state.html.markdown:12]: Clarifying the destroy behavior might be worthwhile here, just to set expectations that state won't change as part of a destroy operation.

```suggestion

~> Destruction of this resource is a no-op and **will not** modify the instance state

```

File: internal/service/rds/rds_instance_state.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+		Attributes: map[string]schema.Attribute{
+			names.AttrIdentifier: schema.StringAttribute{
+				Required: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			names.AttrState: schema.StringAttribute{
+				Required: true,
+			},
```
Comment (jar-b) [inline internal/service/rds/rds_instance_state.go:62]: Might want to make this an enum type, or add a validator to check this is one of `stopped` or `available` ( `updateInstanceState` only handles those values).

File: internal/service/rds/instance.go [hunk 5]
```diff
@@ -3048,3 +3085,43 @@ func flattenEndpoint(apiObject *types.Endpoint) map[string]interface{} {
...
+
+	tflog.Info(ctx, "Starting RDS Instance", map[string]any{
+		"rds_instance_id": id,
+	})
+	_, err = conn.StartDBInstance(ctx, &rds.StartDBInstanceInput{
+		DBInstanceIdentifier: &id,
+	})
+
+	if err != nil {
+		return fmt.Errorf("starting RDS Instance (%s) start: %w", id, err)
```
Comment (jar-b) [inline internal/service/rds/instance.go:3100]:
```suggestion
		return fmt.Errorf("starting RDS Instance (%s): %w", id, err)
```

File: internal/service/rds/instance.go [hunk 6]
```diff
@@ -3048,3 +3085,43 @@ func flattenEndpoint(apiObject *types.Endpoint) map[string]interface{} {
...
+	}
+
+	if _, err := waitDBInstanceAvailable(ctx, conn, id, timeout); err != nil {
+		return fmt.Errorf("waiting for RDS Instance (%s) start %w", id, err)
```
Comment (jar-b) [inline internal/service/rds/instance.go:3104]:
```suggestion
		return fmt.Errorf("waiting for RDS Instance (%s) start: %w", id, err)
```

**Review by jar-b — COMMENTED** [review]
A couple minor suggestions, but I really like the simplicity of the implementation 👍

**Review by jar-b — APPROVED** [review]

**Pull request merged**